File: Filter/Filter.php
<?php

/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/19
 * Time: 20:15
 */
class Person
{
    private $name, $gender, $maritalStatus;

    public function __construct($name, $gender, $maritalStatus)
    {
        $this->name = $name;
        $this->gender = $gender;
        $this->maritalStatus = $maritalStatus;
    }

    public function getName()
    {
        return $this->name;
    }

    public function getGender()
    {
        return $this->gender;
    }

    public function getMaritalStatus()
    {
        return $this->maritalStatus;
    }
}


interface Criteria
{
    public function meetCriteria($persons);
}

class CriteriaMale implements Criteria
{
    public function meetCriteria($persons)
    {
        $malePersons = array();
        // TODO: Implement meetCriteria() method.
        foreach ($persons as $person) {
            $gender = $person->getGender();
            $gender = strtolower($gender);
            if ($gender == "male") {
                array_push($malePersons, $person);
            }
        }
        return $malePersons;
    }

}


class CriteriaFemale implements Criteria
{
    public function meetCriteria($persons)
    {
        // TODO: Implement meetCriteria() method.
        $femalePersons = array();
        foreach ($persons as $person) {
            $gender = $person->getGender();
            $gender = strtolower($gender);
            if ($gender == "female") {
                array_push($femalePersons, $person);
            }
        }

        return $femalePersons;
    }
}


class CriteriaSingle implements Criteria
{
    public function meetCriteria($persons)
    {
        // TODO: Implement meetCriteria() method.
        $singlePersons = array();
        foreach ($persons as $person) {
            $maritalStatus = strtolower($person->getMaritalStatus());
            if ($maritalStatus == "single") {
                array_push($singlePersons, $person);
            }
        }
        return $singlePersons;
    }
}

class AndCriteria implements Criteria
{
    private $criteria;
    private $otherCriteria;

    public function __construct($criteria, $otherCriteria)
    {
        $this->criteria = $criteria;
        $this->otherCriteria = $otherCriteria;
    }

    public function meetCriteria($persons)
    {
        // TODO: Implement meetCriteria() method.
        $firstCriteriaPersons = $this->criteria->meetCriteria($persons);
        return $this->otherCriteria->meetCriteria($firstCriteriaPersons);
    }
}

class OrCriteria implements Criteria
{
    private $criteria;
    private $otherCriteria;

    public function __construct($criteria, $otherCriteria)
    {
        $this->criteria = $criteria;
        $this->otherCriteria = $otherCriteria;
    }

    public function meetCriteria($persons)
    {
        // TODO: Implement meetCriteria() method.
        $firstCriteriaItems = $this->criteria->meetCriteria($persons);
        $otherCriteriaItem = $this->criteria->meetCriteria($persons);
        foreach ($otherCriteriaItem as $person) {
            if (!in_array($person, $firstCriteriaItems)) {
                array_push($firstCriteriaItems, $person);
            }
        }
        return $firstCriteriaItems;
    }
}

function printPersons($persons)
{
    foreach ($persons as $person) {
        echo "Person : [ Name : " . $person->getName()
            . ", Gender : " . $person->getGender()
            . ", Marital Status : " . $person->getMaritalStatus()
            . " ]<br/>";
    }
}

//demo
$persons = array();
array_push($persons, new Person("Robert", "Male", "Single"));
array_push($persons, new Person("John", "Male", "Married"));
array_push($persons, new Person("Laura", "Female", "Married"));
array_push($persons, new Person("Diana", "Female", "Single"));
array_push($persons, new Person("Mike", "Male", "Single"));
array_push($persons, new Person("Bobby", "Male", "Single"));

$male = new CriteriaMale();
$female = new CriteriaFemale();
$single = new CriteriaSingle();
$singleMale = new AndCriteria($single, $male);
$singleOrFemale = new OrCriteria($single, $female);

var_dump("Males: ");
printPersons($male->meetCriteria($persons));

var_dump("Females: ");
printPersons($female->meetCriteria($persons));

var_dump("Single Males: ");
printPersons($singleMale->meetCriteria($persons));

var_dump("Single Or Females: ");
printPersons($singleOrFemale->meetCriteria($persons));

File: Filter/Filter.go
package main

import (
	"container/list"
	"strings"
	"fmt"
)

type Person struct {
	name          string
	gender        string
	maritalStatus string
}

func (p *Person)getName() string {
	return p.name
}

func (p *Person)getGender() string {
	return p.gender
}

func (p *Person)getMaritalStatus() string {
	return p.maritalStatus
}

type Criteria interface {
	meetCriteria(persons list.List) list.List
}

type CriteriaMale struct {

}

func (c *CriteriaMale)meetCriteria(persons list.List) list.List {
	var malePersons list.List
	for p := persons.Front(); p != nil; p = p.Next() {
		person := p.Value.(*Person)
		gender := strings.ToLower(person.getGender())
		if (gender == "male") {
			malePersons.PushBack(person)
		}
	}
	return malePersons
}

type CriteriaFemale struct {

}

func (c *CriteriaFemale)meetCriteria(persons list.List) list.List {
	var femalePersons list.List
	for p := persons.Front(); p != nil; p = p.Next() {
		person := p.Value.(*Person)
		gender := strings.ToLower(person.getGender())
		if (gender == "female") {
			femalePersons.PushBack(person)
		}
	}
	return femalePersons
}

type CriteriaSingle struct {

}

func (c *CriteriaSingle)meetCriteria(persons list.List) list.List {
	var singlePersons list.List
	for p := persons.Front(); p != nil; p = p.Next() {
		person := p.Value.(*Person)
		maritalStatus := strings.ToLower(person.getMaritalStatus())
		if (maritalStatus == "single") {
			singlePersons.PushBack(person)
		}
	}
	return singlePersons
}

type AndCriteria struct {
	criteria      Criteria
	otherCriteria Criteria
}

func (a *AndCriteria)meetCriteria(persons list.List) list.List {
	firstCriteriaPersons := a.criteria.meetCriteria(persons)
	return a.otherCriteria.meetCriteria(firstCriteriaPersons)
}

type OrCriteria struct {
	criteria      Criteria
	otherCriteria Criteria
}

func (o *OrCriteria)meetCriteria(persons list.List) list.List {
	firstCriteriaPersons := o.criteria.meetCriteria(persons)
	otherCriteriaPersons := o.otherCriteria.meetCriteria(persons)
	for p := otherCriteriaPersons.Front(); p != nil; p = p.Next() {
		person := p.Value.(*Person)
		contains, _ := Contains(firstCriteriaPersons, person)
		if (!contains) {
			firstCriteriaPersons.PushBack(person)
		}

	}
	return firstCriteriaPersons
}

func Contains(l list.List, value *Person) (bool, *list.Element) {
	for e := l.Front(); e != nil; e = e.Next() {
		if (e.Value == value) {
			return true, e
		}
	}
	return false, nil
}

func printPersons(p list.List) {
	for e := p.Front(); e != nil; e = e.Next() {
		person := e.Value.(*Person)
		fmt.Println("Person : [ Name : ", person.getName(), ", Gender : ", person.getGender(), ", Marital Status : ", person.getMaritalStatus(), " ]")
	}
}

func main() {
	//demo
	var persons list.List;
	persons.PushBack(&Person{"Robert", "Male", "Single"})
	persons.PushBack(&Person{"John", "Male", "Married"})
	persons.PushBack(&Person{"Laura", "Female", "Married"})
	persons.PushBack(&Person{"Diana", "Female", "Single"})
	persons.PushBack(&Person{"Mike", "Male", "Single"})
	persons.PushBack(&Person{"Bobby", "Male", "Single"})

	male := new(CriteriaMale)
	fmt.Println("Males: ")
	printPersons(male.meetCriteria(persons))

	female := new(CriteriaFemale)
	fmt.Println("Females: ")
	printPersons(female.meetCriteria(persons))

	single := new(CriteriaSingle)

	singleMale := &AndCriteria{single, male}
	fmt.Println("singleMales: ")
	printPersons(singleMale.meetCriteria(persons))

	singleOrFemale := &OrCriteria{single, female}
	fmt.Println("single or females: ")
	printPersons(singleOrFemale.meetCriteria(persons))
}


